cinema/views.py

```diff
@@ -45,9 +76,28 @@ def get_serializer_class(self):
 
 
 class MovieSessionViewSet(viewsets.ModelViewSet):
-    queryset = MovieSession.objects.all()
+    queryset = MovieSession.objects.select_related("movie", "cinema_hall")
     serializer_class = MovieSessionSerializer
 
+    def get_queryset(self):
+        queryset = MovieSession.objects.select_related("movie", "cinema_hall")
+        if self.action == "list":
+            queryset = queryset.annotate(
+                tickets_available=F("cinema_hall__rows")
+                * F("cinema_hall__seats_in_row")
+                - Count("tickets")
+            )
+            movie = self.request.query_params.get("movie")
+            date = self.request.query_params.get("date")
+            if movie:
+                queryset = queryset.filter(movie_id=movie)
+            if date:
+                date = datetime.strptime(date, "%Y-%m-%d").date()
```

Comment:
You don't need to do this, you can simply use retrieved data from query params with transform to date object